### PR TITLE
add requester

### PR DESCRIPTION
**Pull request by pokassun**
- [x] program: create add/remove requester instruction
- [x] program: create requester state
- [x] program: create AddRequesterContext
- [x] cli: add request round
- [x] deployer: add requester configuration
- [x] schema: add requester schema and change AggregatorConfig

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -72,6 +72,14 @@ info:     update {"description":"btc:usd","decimals":2,"roundID":"22","median":"
 info:     update {"description":"btc:usd","decimals":2,"roundID":"23","median":"5740350","updatedAt":"37820565","createdAt":"37820565"}
 ```
 
+# Request New Round
+
+Request a new round for a specific aggregator
+
+```
+yarn solink request-round <aggregator-id>
```
Comment (defactojob) [inline README.md:80]: is aggregator-id the public key or a name like `btc:usd`?

File: deploy.local.json [hunk 2]
```diff
@@ -0,0 +1,102 @@
+{
```
Comment (defactojob) [inline deploy.local.json]: let's not check in `deploy.local.json`

File: program/src/processor.rs [hunk 3]
```diff
@@ -153,6 +153,106 @@ impl<'a> RemoveOracleContext<'a> {
...
+    description: [u8; 32],
+}
+
+impl<'a> AddRequesterContext<'a> {
+    fn process(&self) -> ProgramResult {
+        let aggregator = Aggregator::load_initialized(self.aggregator)?;
+        msg!("loaded aggregator");
+        aggregator.authorize(self.aggregator_owner)?;
+
+        let mut requester = Oracle::init_uninitialized(self.requester)?;
```
Comment (defactojob) [inline program/src/processor.rs]: why is the requester using the same struct as Oracle?

Comment (defactojob) [reply program/src/processor.rs]: i think this is a security issue. If a requester has exactly the same struct as an Oracle, the requester would be able to submit value as an oracle.

Note also that if you create a Requester struct, but its binary representation can deserialize as an Oracle, you'd have the same security hole.

File: program/src/processor.rs [hunk 4]
```diff
@@ -153,6 +153,106 @@ impl<'a> RemoveOracleContext<'a> {
...
+        aggregator.authorize(self.aggregator_owner)?;
+
+        let requester = Requester::load_initialized(self.requester)?;
+        if requester.aggregator.0 != self.aggregator.key.to_bytes() {
+            return Err(Error::AggregatorMismatch)?;
+        }
+
+        // Zero out the requester account memory. This allows reuse or reclaim.
+        // Note: will wipe out withdrawable balance on this requester. Too bad.
+        Oracle::default().save(self.requester)?;
```
Comment (defactojob) [inline program/src/processor.rs]: again, don't use `Oracle` here. very confusing.

File: program/src/state.rs [hunk 5]
```diff
@@ -245,6 +248,48 @@ impl IsInitialized for Oracle {
...
 
+/// Requester data.
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, BorshSchema, Default, PartialEq)]
+pub struct Requester {
+    /// is usually the oracle name
+    pub description: [u8; 32],
+    /// is initialized
+    pub is_initialized: bool,
+    /// TODO: do we need this withdrawable
+    pub withdrawable: u64,
```
Comment (defactojob) [inline program/src/state.rs]: remove withdrawable.

there is a subtle security issue where Requester & Oracle are binary compatible, and allowing either roles to act as the other.

removing `withdrawable` so Requester & Oracle are not binary compatible.

File: src/PriceFeeder.ts [hunk 6]
```diff
@@ -76,5 +80,9 @@ export class PriceFeeder {
 
       submitter.start()
     }
+
+    if(!nFound) {
+      log.error('no oracle found')
```
Comment (defactojob) [inline src/PriceFeeder.ts]: "no matching aggregator to act as oracle"

File: src/RequestFeeder.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
+import { Wallet } from "solray";
```
Comment (defactojob) [inline src/RequestFeeder.ts]: RequestFeeder seems like a bad class name. It's not a "feeder".

Maybe just call it Requester or RoundRequester?

**Review by defactojob — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (pokassun) [reply README.md:80]: is the pair name like `btc:usd`, I will add an example

**Pull request merged**